apollo-router/src/services/layers/persisted_queries/manifest_poller.rs

```diff
@@ -18,102 +21,322 @@ use crate::uplink::persisted_queries_manifest_stream::PersistedQueriesManifestCh
 use crate::uplink::persisted_queries_manifest_stream::PersistedQueriesManifestQuery;
 use crate::uplink::stream_from_uplink;
 use crate::uplink::UplinkConfig;
+use crate::Configuration;
 
 /// An in memory cache of persisted queries.
 pub(crate) type PersistedQueryManifest = HashMap<String, String>;
 
-/// An in memory cache of persisted query bodies.
-pub(crate) type PersistedQuerySet = HashSet<String>;
+/// How the router should respond to requests that are not resolved as the IDs
+/// of an operation in the manifest. (For the most part this means "requests
+/// sent as freeform GraphQL", though it also includes requests sent as an ID
+/// that is not found in the PQ manifest but is found in the APQ cache; because
+/// you cannot combine APQs with safelisting, this is only relevant in "allow
+/// all" and "log unknown" modes.)
+#[derive(Debug)]
+pub(crate) enum FreeformGraphQLBehavior {
+    AllowAll {
+        apq_enabled: bool,
+    },
+    DenyAll {
+        log_unknown: bool,
+    },
+    AllowIfInSafelist {
+        safelist: FreeformGraphQLSafelist,
+        log_unknown: bool,
+    },
+    LogUnlessInSafelist {
+        safelist: FreeformGraphQLSafelist,
+        apq_enabled: bool,
+    },
+}
+
+/// Describes what the router should do for a given request: allow it, deny it
+/// with an error, or allow it but log the operation as unknown.
```

Comment:
Nit. The comment appears to have drifted out of date with the implementation. For this kind of reason, I prefer comments like this:

```
/// How should we treat Freeform GraphQL
pub(crate) enum FreeformGraphQLAction {
    /// Allow it
    Allow,
    /// Deny it, with an error
    Deny,
    /// Allow it, but log the operation as unknown
    AllowAndLog,
    /// Deny it, with an error and log the operation as unknown <- Presumably...
    DenyAndLog,
}
```

Reply:
A durable and defensive suggetion!  Love it.

Reply:
Thanks @BrynCooke for merging! Should I make a follow-up with this comment improvement?

Reply:
If you don't mind.